types/time.go

```diff
@@ -653,21 +654,29 @@ func parseDatetime(sc *stmtctx.StatementContext, str string, fsp int, isFloat bo
 				// 20170118.123423 => 2017-01-18 00:00:00
 			} else {
 				// '20170118.123423' => 2017-01-18 12:34:23.234
-				fmt.Sscanf(fracStr, "%2d%2d%2d", &hour, &minute, &second)
+				switch len(fracStr) {
+				case 0:
+				case 1, 2:
+					_, err = fmt.Sscanf(fracStr, "%2d ", &hour)
+				case 3, 4:
+					_, err = fmt.Sscanf(fracStr, "%2d%2d ", &hour, &minute)
+				default:
+					_, err = fmt.Sscanf(fracStr, "%2d%2d%2d ", &hour, &minute, &second)
+				}
+				truncatedOrIncorrect = err != nil
 			}
 		}
 		if l == 9 || l == 10 {
-			switch len(fracStr) {
-			case 0:
+			if len(fracStr) == 0 {
 				second = 0
-			case 1:
-				_, err = fmt.Sscanf(fracStr, "%1d", &second)
-			case 2:
-				_, err = fmt.Sscanf(fracStr, "%2d", &second)
-			default:
-				_, err = fmt.Sscanf(fracStr[:2], "%2d", &second)
-				sc.AppendWarning(ErrTruncatedWrongValue.GenWithStackByArgs("datetime", str))
+			} else {
+				_, err = fmt.Sscanf(fracStr, "%2d ", &second)
 			}
+			truncatedOrIncorrect = err != nil || len(fracStr) > 2
```

Comment:
Should we leave a comment here for explaining the check `len(fracStr)`?

Reply:
ok